Date controls updates

mark-tate opened the pull request.

mark-tate committed:
`AriaAnnouncer` prop `delay` is deprecated in favour of an `options` prop.

mark-tate committed:
update date input

mark-tate committed:
update calendar

mark-tate committed:
update date picker

mark-tate committed:
change DateFrameworkTypeMap to an interface

mark-tate committed:
biome was breaking the DateFrameworkTypeMap, ignore from biome

mark-tate committed:
try making adapters a peer dependency

mark-tate committed:
fix types

mark-tate committed:
simplify types now that the map is working

mark-tate committed:
make generics optional on remaining types

mark-tate committed:
add changeset

mark-tate committed:
removed unused `getDayOfWeek` from adapters

mark-tate committed:
improve docs
- show errors in Formfield based examples
- remove dependency on getDayWeekName from adapters
- document toJSDate in adapters
- re-enforce the need for LocalizationProvider through examples
- show how to set the date

mark-tate committed:
comment toJSDate

mark-tate committed:
fix tests

mark-tate committed:
Update salty-humans-build.md

fixup changeset

wemmyo commented on .changeset/curly-planes-remain.md:
````diff
@@ -0,0 +1,70 @@
+```
+
+- `CalendarNavigation` now adds tooltips to out of range dates.
+- When a Calendar has a min/maxDate and the user navigates to a month outside of that range, the Calendar will now automatically navigate back to the closest valid month.
````
```suggestion
- When a Calendar has a `minDate` or `maxDate` and the user navigates to a month outside of that range, the Calendar will now automatically navigate back to the closest valid month.
```

tycup commented on packages/core/src/aria-announcer/AriaAnnouncerProvider.tsx:
```diff
@@ -1,82 +1,119 @@
+  const value = useMemo(() => ({ announce }), [announce]);
+
+  const handleMounted = useCallback((node: HTMLDivElement | null) => {
+    mountedRef.current = node !== null;
```
What's `mountedRef.current` used for? I don't see it used as a guard anywhere

tycup commented on packages/core/src/aria-announcer/AriaAnnouncerProvider.tsx:
```diff
@@ -1,82 +1,119 @@
-      });
-    }
-  }, []);
+  const makeAnnouncement = useCallback(
```
I think the legacy delay is being applied twice - once in`useAriaAnnouncer`, via setTimeout and again via `AriaAnnouncerProvider.announce`. Potential fix could be to forward `options` instead of `legacyDelay` after the hook waited

mark-tate commented:
> Looks like we've introduced a visual regression when date picker calendar icon is clicked and the dialog is open, the icon has a grey background

Intended, part of Button

/* Styles applied when the button triggers a dialog or menu */
.saltButton[aria-expanded="true"][aria-haspopup="menu"],
.saltButton[aria-expanded="true"][aria-haspopup="dialog"] {
  background: var(--saltButton-background-active, var(--button-background-active));
  color: var(--saltButton-text-color-active, var(--button-text-color-active));
  border-color: var(--saltButton-borderColor-active, var(--button-borderColor-active));
}

mark-tate committed:
code-review: polish changeset

mark-tate committed:
code-review: cleanup the implementation and avoid case of double announcement

mark-tate committed:
code-review: Announcer used to default to assertive but now defaults to polite
Remove code that preserved assertive and rely on default, polite.

mark-tate committed:
code-review: fix issues with date input labels referencing themselves and make
aria labelled by and label mutually exclusive.

mark-tate committed:
code-review:
- fix issue with console log appearing with duplicate key
- simplify the live region for maximum compatibility

mark-tate committed:
code-review: add aria-multiselectable to CalendarGrid

mark-tate committed:
code-review: remove unused hoveredDate and selectedDate from useCalendar

mark-tate committed:
code-review: add selectedStart, selectedSpan, selectedEnd to useCalendarDay, so we can support a11y toggle buttons for date ranges dates

mark-tate committed:
code-review: add aria-selected to selected dates

mark-tate committed:
code-review: ensure latest references to createAnnouncement are used in announce

mark-tate committed:
code-review: add AriaAnnouncer to overlay as aria-model, ignore live announcements, outside of dialog

mark-tate committed:
code-review: fix createAnnouncement precedence for selectionVariant

mark-tate committed:
code-review:  cleanup calendar stories
- switch to aria-presed for toggled date examples
- cleanup import

mark-tate committed:
core-review: run prettier

mark-tate committed:
core-review: fixup yarn.lock after peer dependency change

mark-tate committed:
code-review: fixed Dayjs timezone handling for date input/picker workflows so that user-entered dates are correctly interpreted as midnight in the selected IANA timezone when serializing to ISO (e.g. `America/New_York`, `Asia/Shanghai`).

mark-tate committed:
code-review: fixed a11y checks for Calendar and DatePicker

Please enter the commit message for your changes. Lines starting

mark-tate committed:
code-review: correct ContentStatus announcement tests

mark-tate committed:
code-review: backed out the mutual exclusive a11y recommendation as it breaks axe

mark-tate committed:
code-review: backed out the use of aria-selected as it breaks axe, use label instead

mark-tate committed:
code-review: prettier/lint

ran lint

mark-tate committed:
code-review: fix content status by making it default to polite

mark-tate committed:
code-review: fixed announcements triggered during component unmount/cleanup (e.g. `Spinner` `completionAnnouncement`) being blocked.

mark-tate committed:
code-review: pip date framework peerDependencies to latest versions, non-breaking

mark-tate committed:
code-review: shorten gap between calendar dropdowns to avoid ellipsis

mark-tate committed:
package-rename: init commit (#6179)

mark-tate committed:
code-review: fix changeset types

mark-tate committed:
cleanup changesets

mark-tate committed:
date adapter comment

mark-tate committed:
code-review: revert assertive default on Spinner and Tooltip, defaults to polite

mark-tate committed:
code-review: improved date picker tests to cover all adapters

mark-tate committed:
code-review: add timezone and clone tests to date-adapters

mark-tate committed:
code-review: fixed date-fns setTimezone so it prserves the wall-clock time as per other adapters

mark-tate committed:
code-review: ensure luxon adapter clone method preserves timezone

mark-tate committed:
code-review: update versions of date frameworks in docs

mark-tate committed:
code-review: update timezone site examples to work with date-fns-tz adapter

mark-tate committed:
code-review: adapter bux fixes
- correct default format for luxon and date-fns
- dayjs does not provide single digit day of week names, so added workaround

mark-tate committed:
code-review: fix DatePicker calendar visible month when user enters out of range date
-

mark-tate committed:
code-review: remove deprecated `DatePickerSinglePanel`

mark-tate committed:
code-review: fixed moment adapter `setTimezone` for UTC to preserve wall-clock time, consistent with all other adapters. Previously, converting a local date to UTC would shift the instant instead of reinterpreting the wall-clock time in UTC.

mark-tate committed:
code-review: ensure Date Picker range examples all use DatePickerRangePanel to show consistency

mark-tate committed:
code-review: fix the error type for defaultRangeValidator in DatePickerRangeInput

mark-tate committed:
code-review: updated spacing on calendar

mark-tate committed:
code-review: set a max-width for CalendarNavigation for grid layout

jake-costa commented:
- [ ] A few DatePicker examples with a visible label provided by a parent FormField do not properly programmatically associate the label with the input, so the visible label is not announced to screen readers. **Fixed Y/N/Comment:**  
  Current NVDA output: date input, edit, date format DD MMM YYYY (e.g. 09 June 2024), blank  
  Expected: Select a date, edit, date format DD MMM YYYY (e.g. 09 June 2024), blank  
  Notes:  
  From what I can tell, this appears to be because aria-labelledby is being added to the wrapper <div> instead of the input.  
  aria-labelledby should be on the <input>, not the wrapper.  
  An input should contain either aria-label or aria-labelledby.  
  
  **Fixed Y/N ?**

- [ ] A few range DatePicker examples with a visible group label do not differentiate the start and end inputs. Both inputs announce the same accessible name, so users cannot tell which input has focus. **Fixed Y/N/Comment:**  
  Current for both inputs: select a date range, edit, dd mmm yyyy, blank  
  Expected for the start input: select a date range, start date, edit, dd mmm yyyy, blank  
  Expected for the end input: select a date range, end date, edit, dd mmm yyyy, blank  
  Note: The default range example (no visible group label) already differentiates start and end correctly via aria-label. This may be a use case for the aria-labelledby self-reference technique when a visible group label is present.  
  
  **Fixed Y/N ?**

- [ ] The month combobox announces the abbreviated month (Jan) instead of the full name when collapsed, and the aria-label duplicates the combobox role by including the word "Dropdown". **Fixed Y/N/Comment:**  
  Current: Month dropdown, combo box, Jan, collapsed  
  Expected: Month, January, combo box, collapsed  
  Note: The same role-duplication applies to the year combobox.  
  
  Cannot be fixed at present.

- [ ] When two calendars are displayed in a range DatePicker, both pairs of previous/next month buttons share the same accessible name (Previous Month, Next Month) with no indication of which calendar they control. Screen reader users navigating via the elements list, and voice-control users speaking the button name, cannot distinguish the four buttons. **Fixed Y/N/Comment:**  

  **Fixed Y/N ?**

  Notes:  
  Preferred fix: collapse to a single shared pair, as shipped by Adobe React Aria RangeCalendar and MUI X DateRangeCalendar. Place Previous Month at the left edge of the first calendar and Next Month at the right edge of the second calendar, and remove the inner pair between the two grids. Update the live-region announcement on navigation to convey both visible months (for example, April 2026 and May 2026).  
  If two pairs are preserved, each pair needs a unique accessible name identifying the calendar it sits on rather than the month it navigates to (for example, Previous Month, April 2026 on the left pair and Previous Month, May 2026 on the right).

- [ ] Add aria-hidden="true" to the decorative SVG icon next to the error message in the initial-error example. 

  **Fixed Y/N ?**
  Not a Date Picker issue, comes from StatusIndicator.

- [ ] After selecting the start day in a range DatePicker, no prompt is announced to select the end date. The date confirmation fires but the guidance prompt is missing. **Fixed Y/N/Comment:**  
  Current: start date. Tuesday 14 April 2026 selected  
  Expected: Start date Tuesday 14 April 2026 selected. Select end date.  
  
  **Fixed Y/N ?**

- [ ] After the start date is selected in a range DatePicker, the remaining day buttons announce complete new range for each candidate end date, and the selected start day's label needs to be modified as well. **Fixed Y/N/Comment:**  
  Current for a candidate end date: complete new range: Wednesday 15 April 2026, button  
  Expected for a candidate end date: Wednesday 15 April 2026, select as end date, button  
  Current for the selected start date: start range: Tuesday 14 April 2026 selected  
  Expected for the selected start date: start date: Tuesday 14 April 2026, selected, button  
 
  **Fixed Y/N ?**

- [ ] Dial back the verbosity of the previous/next month announcement. **Fixed Y/N/Comment:**  
  Current: Visible Month starts from Friday 1 May 2026  
  Expected: May 2026.  
  Note: The W3C APG reference implementations for both the Date Picker Dialog and Date Picker Combobox consistently use the full Month Year format on every navigation, regardless of whether the year has changed.  
 
  **Fixed Y/N ?**

- [ ] At 320×256 px the range DatePicker overlay renders two calendars side by side and overflows the viewport. Per dev team this is consumer-implementation responsibility; flagging for potential documentation update. **Fixed Y/N/Comment:**  
 
   **No action planned**

mark-tate committed:
a11y-review: fix aria-labels for date-inputs

mark-tate committed:
a11y-review: ensure DateInputRange and FormField announces whether it is start or end date

mark-tate committed:
a11y-review: ensure the DateRangePicker next/previous month buttons have different labels

mark-tate committed:
a11y-review: fixed range DatePicker screen reader guidance so selecting a start date now announces a follow-up prompt to choose the end date

mark-tate committed:
a11y-review: improve the date range completion announcements

mark-tate committed:
a11y-review: dial back the verbosity of the previous/next month announcement.

mark-tate committed:
a11y-review: fix tests

jake-costa commented:
## Accessibility Findings

* The month combobox announces the abbreviated month (`Jan`) instead of the full name when collapsed, and the `aria-label` duplicates the combobox role by including the word "Dropdown".

  * Current: `Month dropdown, combo box, Jan, collapsed`
  * Expected: `Month, January, combo box, collapsed`
  * Note: The same role-duplication applies to the year combobox.

* The decorative SVG icon next to the error message in the initial-error example is still announced. Add `aria-hidden="true"` so screen readers skip it, since the error state is already conveyed through the helper text and `aria-invalid` on the input.

* After selecting the start day in a range `DatePicker`, no prompt is announced to select the end date. The date confirmation fires but the guidance prompt is missing.

  * Current: `start date. Tuesday 14 April 2026 selected`
  * Expected: `start date. Tuesday 14 April 2026 selected.` followed by `Select end date.`
  * Note: The start-date confirmation already announces correctly via the day button's accessible name, only the `Select end date.` prompt needs to fire from the live region on top of it. Putting the full date back into the live-region message would double up on what the button already announces.

* A few of the locale examples are missing the `lang` attribute on the wrapping element. Some examples do this correctly, but `DateInput`'s `SingleWithLocale` and `RangeWithLocale` still need it.

* In the custom-content example, the listbox is not implemented correctly, its options have `role="button"` instead of `role="option"`, and navigating between the options with the screen reader does not work. Initial focus on open also lands in the current day in the calendar rather than on the first option in the listbox.

  * Note: This may be specific to the example rather than the underlying `ListBox` component, but worth flagging since consumers may copy the pattern.
  
* Lastly, both `<h4 class="saltText saltText-primary">Note</h4>` on the example page looks to be getting flagged by aXe. (Should likely remove heading semantics)

jake-costa commented:
## Accessibility Findings

### Priority

* The month combobox announces the abbreviated month (`Jan`) instead of the full name when collapsed, and the `aria-label` duplicates the combobox role by including the word "Dropdown".

  * Current: `Month dropdown, combo box, Jan, collapsed`
  * Expected: `Month, January, combo box, collapsed`
  * Note: The same role-duplication applies to the year combobox.

### Low priority

* The decorative SVG icon is still not properly hidden.

### Usability enhancements (Not a WCAG violation)

* When the month or year is changed using the Month or Year dropdown, the live region also announces the new month and year (for example, `May 2026.`) on top of the dropdown's own selection feedback. The dropdown already conveys the selected value to AT, so the live-region message is not needed.

* After selecting the start day in a range `DatePicker`, no follow-up guidance is announced to direct the user to select the end date.
  * Recommended fix: announce exactly `Select end date.` from the live region after start-only selection. Do not repeat the date or the word "selected" in the live-region string, since the button's accessible name already covers that and including it would double up.

## Example Cleanup

* A few of the locale examples are missing the `lang` attribute on the wrapping element. Most examples do this correctly, but `DateInput`'s `SingleWithLocale` and `RangeWithLocale` still need it. Setting the `lang` attribute is typically a consumer responsibility in the host HTML and the `LocalizationProvider` does not emit it, so this is not a component defect. Flagging for example consistency, since other locale examples (`SingleWithLocaleZhCN`) do set it and the inconsistency could mislead downstream teams copying the pattern.

* In the custom-content example, the listbox is not implemented correctly. Its options have `role="button"` instead of `role="option"`, and navigating between the options with the keyboard is not possible.

mark-tate committed:
a11y-review: add lang to date-input locale examples

mark-tate committed:
a11y-review: improved incomplete range announcements

mark-tate committed:
a11y-review: correct role on custom panel example

mark-tate committed:
a11y-review: fix duplicate announcements

mark-tate commented:
/release-pr

tycup commented on packages/date-adapters/src/date-fns-tz-adapter/index.ts:
```diff
@@ -66,26 +84,44 @@ export class AdapterDateFnsTZ extends AdapterDateFns {
+    // We extract the local date/time components from the source date and recreate them
+    // in the target timezone so that "Jan 5 midnight" stays "Jan 5 midnight" regardless
+    // of which timezone is selected.
+    return new TZDate(
```
does this work with strings that encode an instant (and does this matter here?)

naive date strings work fine, but strings like "2025-01-05T00:00:00.000Z" will produce different results depending on the host timezone. hose getters read in the machine's local zone before the value is reinterpreted in the target zone. E.g.: under TZ=Europe/London it resolves to 2025-01-05T00:00:00.000-05:00, while under TZ=America/New_York it resolves to 2025-01-04T19:00:00.000-05:00. Also conflicts with the behaviour the Day.js adapter tests  which now lock in for "…Z" inputs

tycup commented on packages/date-components/src/date-picker/useDatePicker.ts:
```diff
@@ -186,9 +180,18 @@ export function useDatePicker<
   const datePickerRef = useRef<HTMLDivElement>(null);
   const containerRef = useForkRef(ref, datePickerRef);
 
+  const { announce } = useDateSelectionAnnouncer(
```
based on code-review: add AriaAnnouncer to overlay as aria-model commit

if the intention is to move date-picker announcements inside the modal dialog,  useDatePicker() calls useDateSelectionAnnouncer() before DatePickerOverlay renders its new AriaAnnouncerProvider. 

I think that this means the data-selection announcements can continue to go to the live region outside the aria-model dialog (i.e. the new provider is rendered too low in th tree to fix the bug i think you are trying to fix)

mark-tate replied:
fixed

tycup commented on packages/core/src/aria-announcer/AriaAnnounce.tsx:
```diff
@@ -1,24 +1,36 @@
 import { useAriaAnnouncer } from "./useAriaAnnouncer";
 
-export interface AriaAnnounceProps {
+export interface AriaAnnounceProps extends AnnounceFnOptions {
```
AriaAnnounceProps still extends AnnounceFnOptions, so the public API still advertises duration, but the implementation now only forwards { ariaLive }. This means <AriaAnnounce duration={//} /> silently falls back to the provider default instead of the requested time

mark-tate committed:
ai-fix: preserve explicit-offset ISO instants across timezone adapters
- fix `AdapterDateFnsTZ.date` to preserve epoch for ISO values with explicit offset (`Z` or `+/-HH:mm`)
- broaden `AdapterDayjs` explicit-instant handling from `Z`-only to all explicit offsets
- add regression tests for `Z` and numeric-offset instant preservation in:
  - `date-fns-tz.spec.ts`
  - `dayjs.spec.ts`
  - `moment.spec.ts`
  - `luxon.spec.ts`

mark-tate committed:
a11y-review: prevent passive calendar announcements and handle Calendar Today-button programmatic selection
- replace useEffect to trigger announcements by invoking through onSelectionChange

mark-tate committed:
a11y-review: add targeted aria announcer routing and fix DatePicker overlay announcements

- add target-based announcement routing support in aria announcer internals
  (`announcementRegistry`, provider, hook/context updates)
- route DatePicker overlay announcements to an in-dialog announcer target
- restore calendar navigation announcements in DatePicker panels while
  suppressing duplicate dateSelected messages where parent handles selection

mark-tate committed:
a11y-review: honor explicit aria-labels in date inputs and add a11y tests

- DateInputRange: allow explicit top-level/per-input aria-labels to take effect
  by omitting aria-labelledby when explicit labels are provided
- DateInputSingle: apply the same aria-label precedence behavior for consistency

mark-tate committed:
a11y-review: run custom validators for unset values in date picker inputs

DatePickerSingleInput no longer gates custom `validate` behind a truthy `date`,
so consumer validators now run for cleared/null selections as well.

mark-tate committed:
lint

mark-tate committed:
calendar bug fix and remove one useEffect

mark-tate committed:
reset state if for some reason DatePickerActions unmounts

joshwooding commented on package.json:
```diff
@@ -16,8 +16,9 @@
     "build:ag-grid-theme": "yarn workspace @salt-ds/ag-grid-theme build",
     "build:highcharts-theme": "yarn workspace @salt-ds/highcharts-theme build",
-    "bundle:css": "yarn bundle:core:css && yarn bundle:lab:css && yarn bundle:embla-carousel:css && yarn copy:icon:css && yarn copy:countries:css",
+    "bundle:css": "yarn bundle:core:css && yarn bundle:lab:css && yarn bundle:date-components:css & yarn bundle:embla-carousel:css && yarn copy:icon:css && yarn copy:countries:css",
```
```suggestion
    "bundle:css": "yarn bundle:core:css && yarn bundle:lab:css && yarn bundle:date-components:css && yarn bundle:embla-carousel:css && yarn copy:icon:css && yarn copy:countries:css",
```

mark-tate replied:
added on my branch

joshwooding commented on packages/core/src/aria-announcer/useAriaAnnouncer.ts:
```diff
@@ -33,15 +50,28 @@ export const useAriaAnnouncer: useAriaAnnouncerHook = ({
-        if (mountedRef.current) {
-          if (isReactAnnouncerInstalled) {
-            context.announce(announcement);
+        console.log(announcement);
```
```suggestion
```

mark-tate replied:
Removed on my branch

joshwooding commented on .changeset/chalky-white-board.md:
```diff
@@ -0,0 +1,16 @@
+---
+"@salt-ds/lab": patch
```
```suggestion
"@salt-ds/lab": minor
```

mark-tate committed:
ai-review: refactor date selection confirmation, remove effect and replace with a prop `enableApply`

mark-tate committed:
ai-review: disable end calendar Next button based on maxDate instead of minDate

mark-tate committed:
ai-review: fix DatePickerActions labels for undefined selection and partial ranges

joshwooding commented on packages/date-components/src/calendar/CalendarNavigation.css:
```diff
@@ -1,10 +1,13 @@
+  max-width: var(--calendarNavigation-maxWidth, none);
 }
 
 .saltCalendarNavigation-yearDropdown {
```
```suggestion
.saltDropdown.saltCalendarNavigation-yearDropdown {
```

mark-tate replied:
done

joshwooding commented on packages/date-components/src/calendar/CalendarNavigation.css:
```suggestion
  min-width: min-content;
```

mark-tate replied:
done

mark-tate committed:
ai-review: destructure useCalendarSelection callback deps to avoid unnecessary re-creation

mark-tate committed:
fix tests

mark-tate committed:
correct date-components changeset

mark-tate committed:
more specifity for calendar navigation year dropdown

mark-tate committed:
fix docs example for responsive DatePicker to show 2 calendars at xs breakpoint

joshwooding commented on .changeset/chalky-white-board.md:
```diff
@@ -0,0 +1,54 @@
+---
+"@salt-ds/lab": minor
+"@salt-ds/date-components": patch
```
```suggestion
"@salt-ds/date-components": minor
```

joshwooding committed:
Update .changeset/chalky-white-board.md

joshwooding commented on packages/date-components/src/calendar/internal/CalendarDay.css:
```diff
@@ -254,18 +254,18 @@
   position: absolute;
   left: 0;
-  right: 0;
+  right: -1px;
```
Hardcoded value?

mark-tate replied:
agreed to have this hardcoded a while ago.
Can add a token if it needs to change but don't see many folks recreating the calendar grid.

joshwooding commented on packages/date-components/src/localization-provider/LocalizationProvider.tsx:
```diff
@@ -78,8 +82,8 @@ if (process.env.NODE_ENV !== "production") {
 }
 
 export const LocalizationProvider = function LocalizationProvider<
```
Not sure if this should be in date controls? If so maybe we rename it to be more specific?

joshwooding commented:
/release-pr

mark-tate committed:
Merge branch 'main' into screenreader-a11y-changes

joshwooding committed:
Merge branch 'main' into screenreader-a11y-changes
